[Feature] Add action to autorequire undefined globals

sewbacca opened the pull request:
I attempted to implement #1938.
This is how it looks in action:
![autorequire-rewrite](https://github.com/LuaLS/lua-language-server/assets/15840247/f6da7304-63b4-470e-ba5c-af41b3a00606)
How to make this less of a hack?
https://github.com/LuaLS/lua-language-server/blob/62d4d35faf325788355ea9854bee54383db5acdb/script/core/code-action.lua#L730-L739
How to add tests? I suppose they should go somewhere here:
https://github.com/LuaLS/lua-language-server/blob/62d4d35faf325788355ea9854bee54383db5acdb/test/code_action/init.lua#L158
But how do I create the stubs for other modules?

sumneko commented:
> But how do I create the stubs for other modules?

See `test/crossfile`

sewbacca commented:
I added a testcase, and used the `test/crossfile` directory as a reference (though the actual test lies inside the `code_action/init.lua` file). I implemented a custom `TEST_CROSSFILE` function to setup the unit test. I have also modified the eq function, mainly to return a more friendly error message, showing exactly the keypath that does not match.

I also don't use diagnostic messages anymore to determine if a global is undefined, so that I can add the `refactor.rewrite` command. Instead I use the `core.diagnostic.undefined-global` function to report the undefined globals back to me and double check, if the requested position from the user is actually undefined. It still feels a bit of a hack, but it should be much more robust now, since I don't use a diagnostic message as a reference. Is this function called when the user asks for code fixes or more often? Because as far as I understand, the function checks the whole ast, not just only the part in which we are interested. Is there maybe a function, which can check if a given variable is an undefined global directly?

The undefined-global function is marked as async, but the code-action function is not. Can I safely add a `---@async` statement to remove the warning, since the tests pass?

sumneko commented:
> I have also modified the eq function, mainly to return a more friendly error message, showing exactly the keypath that does not match.

Greate, thank you!

> Is this function called when the user asks for code fixes or more often? Because as far as I understand, the function checks the whole ast, not just only the part in which we are interested. Is there maybe a function, which can check if a given variable is an undefined global directly?

There is currently no such function. You can do the following:

Extract the check function in undefined-global into a separate function.
Use `guide.eachSourceBetween` to only check the interested part.
Cache the results in node.

```lua
---@class vm.node
---@field package _undefined_global? boolean

function api.isUndefinedGlobal(src)
    local node = vm.compileNode(src)
    if node._undefined_global == nil then
        node._undefined_global = checkIsUndefinedGlobal(src)
    end
    return node._undefined_global
end
```

> The undefined-global function is marked as async, but the code-action function is not. Can I safely add a `---@async` statement to remove the warning, since the tests pass?

Yes, all code is executed within a coroutine during testing.

sumneko commented:
> > I have also modified the eq function, mainly to return a more friendly error message, showing exactly the keypath that does not match.
> 
> Greate, thank you!
> 
> > Is this function called when the user asks for code fixes or more often? Because as far as I understand, the function checks the whole ast, not just only the part in which we are interested. Is there maybe a function, which can check if a given variable is an undefined global directly?
> 
> There is currently no such function. You can do the following:
> 
> Extract the check function in undefined-global into a separate function. Use `guide.eachSourceBetween` to only check the interested part. Cache the results in node.
> 
> ```lua
> ---@class vm.node
> ---@field package _undefined_global? boolean
> 
> function api.isUndefinedGlobal(src)
>     local node = vm.compileNode(src)
>     if node._undefined_global == nil then
>         node._undefined_global = checkIsUndefinedGlobal(src)
>     end
>     return node._undefined_global
> end
> ```
> 
> > The undefined-global function is marked as async, but the code-action function is not. Can I safely add a `---@async` statement to remove the warning, since the tests pass?
> 
> Yes, all code is executed within a coroutine during testing.

(Translated by ChatGPT)

sewbacca commented:
I've moved the undefined global check into `script/vm` and rewrote `code-action` and `diagnostics/undefined-global` to use the vm instead. I also added another test, if an init file is present. I believe it is no ready for review.

sumneko commented on script/core/code-action.lua:
```diff
@@ -676,6 +681,53 @@ local function checkJsonToLua(results, uri, start, finish)
+        end)
+    end
+
+    guide.eachSourceBetween(state.ast, start, finish, function (source)if vm.isUndefinedGlobal(source) then
```
small formatting issue

sumneko commented on script/vm/global.lua:
```diff
@@ -518,6 +519,33 @@ function vm.hasGlobalSets(suri, cate, name)
+---@return boolean
+function vm.isUndefinedGlobal(src)
+    local node = vm.compileNode(src)
+    if node._undefined_global == nil then
```
My previous example used the wrong style. Please change the fields to lower camel case.

sewbacca commented:
Fixed formatting
renamed `_undefined_global` to `undefinedGlobal`

sumneko commented:
Nice work, thank you!
